source/remoteClient/dialogs.py

```diff
@@ -0,0 +1,336 @@
+		t = threading.Thread(target=self.keyConnector.run)
+		t.start()
+
+	@alwaysCallAfter
+	def handleKeyGenerated(self, key: Optional[str] = None) -> None:
+		self.key.SetValue(key)
+		self.key.SetFocus()
+		self.keyConnector.close()
+		self.keyConnector = None
+
+	@alwaysCallAfter
+	def handleCertificateFailed(self) -> None:
+		try:
+			certHash = self.keyConnector.lastFailFingerprint
+
+			wnd = CertificateUnauthorizedDialog(None, fingerprint=certHash)
+			a = wnd.ShowModal()
+			if a == wx.ID_YES:
+				config = configuration.get_config()
+				config["trusted_certs"][self.host.GetValue()] = certHash
+			if a != wx.ID_YES and a != wx.ID_NO:
+				return
+		except Exception as ex:
+			log.error(ex)
+			return
+		self.keyConnector.close()
+		self.keyConnector = None
+		self.generateKeyCommand(True)
+
+
+class PortCheckResponse(TypedDict):
+	host: str
+	port: int
+	open: bool
+
+
+class ServerPanel(wx.Panel):
+	getIP: wx.Button
+	externalIP: wx.TextCtrl
+	port: wx.TextCtrl
+	key: wx.TextCtrl
+	generateKey: wx.Button
+
+	def __init__(self, parent: Optional[wx.Window] = None, id: int = wx.ID_ANY):
+		super().__init__(parent, id)
+		sizer = wx.BoxSizer(wx.HORIZONTAL)
+		# Translators: Used in server mode to obtain the external IP address for the server (controlled computer) for direct connection.
+		self.getIP = wx.Button(parent=self, label=_("Get External &IP"))
+		self.getIP.Bind(wx.EVT_BUTTON, self.onGetIP)
+		sizer.Add(self.getIP)
+		# Translators: Label of the field displaying the external IP address if using direct (client to server) connection.
+		sizer.Add(wx.StaticText(self, wx.ID_ANY, label=_("&External IP:")))
+		self.externalIP = wx.TextCtrl(self, wx.ID_ANY, style=wx.TE_READONLY | wx.TE_MULTILINE)
+		sizer.Add(self.externalIP)
+		# Translators: The label of an edit field in connect dialog to enter the port the server will listen on.
+		sizer.Add(wx.StaticText(self, wx.ID_ANY, label=_("&Port:")))
+		self.port = wx.TextCtrl(self, wx.ID_ANY, value=str(SERVER_PORT))
+		sizer.Add(self.port)
+		sizer.Add(wx.StaticText(self, wx.ID_ANY, label=_("&Key:")))
+		self.key = wx.TextCtrl(self, wx.ID_ANY)
+		sizer.Add(self.key)
+		self.generateKey = wx.Button(parent=self, label=_("&Generate Key"))
+		self.generateKey.Bind(wx.EVT_BUTTON, self.onGenerateKey)
+		sizer.Add(self.generateKey)
+		self.SetSizerAndFit(sizer)
+
+	def onGenerateKey(self, evt: wx.CommandEvent) -> None:
+		evt.Skip()
+		res = str(random.randrange(1, 9))
+		for n in range(6):
+			res += str(random.randrange(0, 9))
+		self.key.SetValue(res)
+		self.key.SetFocus()
+
+	def onGetIP(self, evt: wx.CommandEvent) -> None:
+		evt.Skip()
+		self.getIP.Enable(False)
+		t = threading.Thread(target=self.doPortcheck, args=[int(self.port.GetValue())])
+		t.daemon = True
+		t.start()
+
+	def doPortcheck(self, port: int) -> None:
+		tempServer = server.LocalRelayServer(port=port, password=None)
+		try:
+			req = request.urlopen("https://portcheck.nvdaremote.com/port/%s" % port)
```

Comment:
If the port checker is under our control, it ideally avoids http, since http can mangle the returned IP when behind a proxy. I'm actually in that situation myself, so port check doesn't give me the proper IP

Reply:
I recommend that we merge the current portcheck functionality and then open a further issue/discussion to explore a non-HTTP-based portcheck service.